### PR TITLE
TLSv1.3: fix padding extension

### PR DESCRIPTION
**Pull request by mattcaswell**
##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

##### Description of change

In OpenSSL 1.1.0 the padding extension MUST be last because it calculates
the length of everything that has been written into the ClientHello to
determine whether it needs to be padded or not. With TLSv1.3 that isn't
possible because the specification requires that the PSK extension is last.
Therefore we need to fix the padding extension to take account of any PSK
extension that will be later added.

### CODE REVIEW

File: test/clienthellotest.c [hunk 1]
```diff
@@ -15,144 +15,217 @@
...
+int test_main(int argc, char *argv[])
+{
+    int testresult;
+
+    ADD_ALL_TESTS(test_client_hello, TOTAL_NUM_TESTS);
+
+    testresult = run_tests(argv[0]);
 
-    return testresult?0:1;
+    return testresult;
```
Comment (levitte) [inline test/clienthellotest.c]: I would remove the use of `testresult` entirely and just:

``` C
    return run_tests(argv[0]);
```

Comment (mattcaswell) [inline test/clienthellotest.c]: Good point. Done.

**Comment by mattcaswell** [comment]
Updated to take account of feedback. Also fixed an issue with out-of-tree builds (could not find the session file)

File: test/clienthellotest.c [hunk 2]
```diff
@@ -15,144 +15,220 @@
...
+    if (!testresult)
+        printf("ClientHello test: FAILED (Test %d)\n", currtest);
+
+    return testresult;
+}
+
+int test_main(int argc, char *argv[])
+{
+    if (argc != 2)
+        return 0;
```
Comment (richsalz) [inline test/clienthellotest.c]: return 1?

Comment (levitte) [inline test/clienthellotest.c]: or `return EXIT_FAILURE`, that will make it easier to understand.

Comment (mattcaswell) [inline test/clienthellotest.c]: Fixed

**Comment by mattcaswell** [comment]
New commit pushed which addresses above feedback. It also fixes a problem where TLSv1.3 is not enabled.

File: ssl/statem/extensions_clnt.c [hunk 3]
```diff
@@ -751,6 +784,12 @@ int tls_construct_ctos_psk(SSL *s, WPACKET *pkt, unsigned int context, X509 *x,
     s->session->ext.tick_identity = TLSEXT_PSK_BAD_IDENTITY;
 
     /*
+     * Note: At this stage we only support adding a single resumption PSK. If
```
Comment (richsalz) [inline ssl/statem/extensions_clnt.c]: at this stage of the code, or the protocol?

Comment (mattcaswell) [inline ssl/statem/extensions_clnt.c]: Fixed.

**Comment by richsalz** [comment]
+1

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**